doc/dev_guide/c_style_guide_new.rst

```diff
@@ -0,0 +1,1046 @@
+next instruction in the assembly output:
+
+.. code-block:: c
+
+  asm ("magic %reg1, #42\n\t"
+       "more_magic %reg2, %reg3"
+       : /* outputs */ : /* inputs */ : /* clobbers */);
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 16: Conditional Compilation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Wherever possible, don't use preprocessor conditionals (#if, #ifdef) in .c
+files; doing so makes code harder to read and logic harder to follow.  Instead,
+use such conditionals in a header file defining functions for use in those .c
+files, providing no-op stub versions in the #else case, and then call those
+functions unconditionally from .c files.  The compiler will avoid generating
+any code for the stub calls, producing identical results, but the logic will
+remain easy to follow.
+
+Prefer to compile out entire functions, rather than portions of functions or
+portions of expressions.  Rather than putting an ifdef in an expression, factor
+out part or all of the expression into a separate helper function and apply the
+conditional to that function.
+
+If you have a function or variable which may potentially go unused in a
+particular configuration, and the compiler would warn about its definition
+going unused, mark the definition as __maybe_unused rather than wrapping it in
+a preprocessor conditional.  (However, if a function or variable *always* goes
+unused, delete it.)
+
+Within code, where possible, use the IS_ENABLED macro to convert a Kconfig
+symbol into a C boolean expression, and use it in a normal C conditional:
+
+.. code-block:: c
+
+  if (IS_ENABLED(CONFIG_SOMETHING)) {
+    ...
+  }
+
+The compiler will constant-fold the conditional away, and include or exclude
+the block of code just as with an #ifdef, so this will not add any runtime
+overhead.
+However, this approach still allows the C compiler to see the code
+inside the block, and check it for correctness (syntax, types, symbol
+references, etc).  Thus, you still have to use an #ifdef if the code inside the
+block references symbols that will not exist if the condition is not met.
+
+At the end of any non-trivial #if or #ifdef block (more than a few lines),
+place a comment after the #endif on the same line, noting the conditional
+expression used.  For instance:
+
+.. code-block:: c
+
+  #ifdef CONFIG_SOMETHING
+  ...
+  #endif /* CONFIG_SOMETHING */
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 17: Header files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. admonition:: Tarantool Style
+    :class: FACT
+
+    Use header guards. Put the header guard in the first line in the header,
+    before the copyright or declarations. Use all-uppercase name for the header
+    guard. Derive the header guard name from the file name, and append _INCLUDED
+    to get a macro name. For example, core/log_io.h -> CORE_LOG_IO_H_INCLUDED. In
+    ``.c`` (implementation) file, include the respective declaration header before
```

Comment:
We use `#pragma once` in the new code instead of macro guards.

Reply:
I need more details to rewrite this part

Reply:
It is that simple - just use `#pragma once` in the headers. If we need to add some water, then we can say that as the header guards we refer to this construction:
```C
#ifndef THE_HEADER_IS_INCLUDED
#define THE_HEADER_IS_INCLUDED

// ... the header code ...

#endif // THE_HEADER_IS_INCLUDED
```
It works fine, but the guard name `THE_HEADER_IS_INCLUDED` tends to outdate when the file is moved or renamed. This is especially painful with multiple files having the same name in the project, but different path. For instance, we have 3 `error.h` files, which means for each of them we need to invent a new header guard name, and not forget to update them if the files are moved or renamed.

For that reason we use `#pragma once` in all the new code, which shortens the header file down to this:
```C
#pragma once

// ... header code ...
```

(Header guard and pragma once - both prevent inclusion of the same file multiple times in C when `#include "<my_file_name>.h"` is used.)